collector/src/main/java/io/prometheus/jmx/JmxCollector.java

```diff
@@ -109,19 +109,18 @@ private void reloadConfig() {
       }
     }
 
-    private void maybeReloadConfig() {
+    private Config maybeReloadConfig() {
       if (configFile != null) {
-        long mtime = configFile.lastModified();
-        if (mtime > config.lastUpdate) {
-          synchronized (this) {
-            // check again in case two threads try to reload the config at the same time
-            if (mtime > config.lastUpdate) {
-              LOGGER.fine("Configuration file changed, reloading...");
-              reloadConfig();
-            }
+        synchronized (this) {
+          long mtime = configFile.lastModified();
+          if (mtime > config.lastUpdate) {
+            LOGGER.fine("Configuration file changed, reloading...");
+            reloadConfig();
           }
         }
       }
+
+      return config;
```

Comment:
This reference to config need to come from inside the lock.

Reply:
Good point. I've made the whole method `synchronized` now.